### ISSUE TITLE
UI for Search, Sort, Filtering, Pagination

### ISSUE THREAD

**Issue by clash99** [Cadasta/cadasta-platform#76]
Reusable components for list views (links to final wireframes).

- [x] [Search](https://drive.google.com/open?id=0BzpiEtMtHC3rVzlzYUhhU09kSmM)
- [x] [Sort](https://drive.google.com/open?id=0BzpiEtMtHC3rM3ZhLWFqQUpGY1E)
- [x] [Filtering](https://drive.google.com/open?id=0BzpiEtMtHC3rREVURGdPeklZREU)
- [x] [Pagination](https://drive.google.com/open?id=0BzpiEtMtHC3raEctZ2tweEN6blE)

**Comment by clash99** [comment]
Wireframe options:

- [Search](https://drive.google.com/open?id=0BzpiEtMtHC3rc3hqR1pyZWVCOWM)
- [Sort](https://drive.google.com/open?id=0BzpiEtMtHC3rc3hqR1pyZWVCOWM)
- [Filter: Option 1](https://drive.google.com/open?id=0BzpiEtMtHC3rc3hqR1pyZWVCOWM)
- [Filter: Option 2](https://drive.google.com/open?id=0BzpiEtMtHC3rU000WC1IOHJoYmM)
- [Filter: Option 3](https://drive.google.com/open?id=0BzpiEtMtHC3rWUptUUxtQlpnbDg)
- [Pagination Options](https://drive.google.com/open?id=0BzpiEtMtHC3rbDlQeV9oX2EtbHM)

**Comment by ian-ross** [comment]
I like option 2 for filtering here.  It's marginally more complex, but it's definitely the most flexible and it's quite in keeping with filtering options on other websites.  Also, I think the point about the filter restricting the area available for data can be addressed by making it possible to collapse the filter area without deactivating the filter: you'd have a different icon to show a filter is active (bright green filter icon vs. grey to make it really obvious?).  That way, you can expand the filter panel to set up a filter, then once you're done, you can collapse the filter panel to browse data.

For pagination, I like option 2, with the addition of "first page" and "last page" buttons around the "previous page" and "next page" buttons.  Being able to enter a page number to jump to directly is very useful once you have a lot of data.  I have one website I use quite a lot that does this very badly, only allowing you to step by a single page or to the last page in a series of 10 or so -- it can take a really long time to find what you want.  Being able to put a page number in, you can often guess more or less where what you want will be if things are sorted alphabetically.

One other possible option is "infinite scrolling" (like in Facebook and Twitter), where you don't really have pagination at the UI level and the user just scrolls to where they want to be with the front end requesting more data from the backend as required for display (this is handled using something like pagination at the API level, though the user doesn't see that).  I don't know how well that would work.  Technically, it's probably not much more difficult than standard pagination.

As well as the static views of the UI layout, it would also be useful to think about what user interactions with these options look like: presumably we want to have dynamic "as you type" searching at least?

@clash99: One other comment.  A few times, you've talked about us needing something "Ajaxy" to do certain things.  The way we're doing the front end, pretty much *everything* is "Ajaxy" -- this really will be a single page application, with *no* page reloads at all.  That means that it's not much more difficult to do really dynamic things than it is to do "traditional" interaction models, so if you ever feel like you should hold back from an idea because of that, don't!

**Comment by wonderchook** [comment]
**Filter**: I like option 3, though I think we'd need to figure out a way to make sure people know what the filter icon is. Just thinking about people that haven't used that many types of software. I think it is easier to line-up which column you are trying to filter with that interface since you are directly interacting at the column header. 

**Pagination**: I like option 2 as well. I think it will work better when there are huge amounts of data. @ian-ross Not sure infinite scrolling would be intuitive in this scenario, but I'd be interested to see examples that use it for tabular data like this. I think with FB and Twitter you aren't generally looking for a specific row, but you might be in our use case.

**Comment by clash99** [comment]
Hi guys - thanks for the feedback.  I've created final versions and linked to them in the checkbox to-do list in the description.

For pagination I finalized option 2.  

For filtering I made a modified version of option 3 for initial implementation and I'm flagging option 2 as more of an advanced option that we can potentially implement in the future.  The changes to option 3 include adding a search box to the filter popover.  We can discuss more in the tech meeting on Wednesday.
